src/Numeric/Units/Dimensional/Dynamic.hs

```diff
@@ -101,19 +101,23 @@ instance Num a => Monoid (AnyQuantity a) where
 -- not be a 'Quantity' of any 'Dimension' whatsoever, but in exchange it gains instances
 -- for the common numeric classes. It's therefore useful for manipulating, and not merely storing,
 -- quantities of unknown dimension.
-newtype DynQuantity a = DynQuantity (Maybe (AnyQuantity a))
-  deriving (Eq, Data, Generic, Generic1, Typeable, Show)
+data DynQuantity a = GoodQuantity (AnyQuantity a) | BadQuantity
```

Comment:
What about going all the way to:
`data DynQuantity a = Good Dimension' a | Bad`

(or `GoodQ`/`BadQ` or `GoodQuantity`/`BadQuantity`, but my point is inlining the fields of `AnyQuantity` since you are always going to match on them).

Another option would be keeping the existing definition and adding bidirectional pattern synonyms:

```
pattern GoodQ d v = DynQuantity (Just (AnyQuantity d v))
pattern BadQ = DynQuantity Nothing
```

in which case you could keep the good parts of the monadic approach, if there are any.